### ISSUE TITLE
Model checkpointing to just save the latest model

### ISSUE THREAD

**Issue by hjalmarlucius** [PyTorchLightning/pytorch-lightning#851]
## 🚀 Feature
There is currently no automated option of always only keeping the latest saved model - there is always a metric benchmark used that needs to be exceeded. This is fine as a default but there should be a 'keep-it-simple' option of always just saving whatever is the latest.

### Motivation

The concept of top-k is sound but when testing the capacity of new architectures, the most important is just to be resume the latest version. This can probably be implemented by using `max(self.global_step)` as the metric - but this seems overly complex.

### Alternatives

An option to always save the latest model and no others - e.g. by setting `save_top_k=-1` (instead of disabling saving altogether which is better done by not having a `ModelCheckpoint`).

**Comment by festeh** [comment]
If one needs to keep just a single checkpoint, it should be the best in some way, say by val accuracy. What is the rationale behind keeping just a last checkpoint?

**Comment by kuynzereb** [comment]
> An option to always save the latest model and no others - e.g. by setting save_top_k=-1 (instead of disabling saving altogether which is better done by not having a ModelCheckpoint).

Actually, if you set `save_top_k=-1` in the current `ModelCheckpoint` it should save every checkpoint. But yes, it is not possible to save **only** the last checkpoint. Moreover, there may be no validation loop in your setup in which case current `ModelCheckpoint` can't be used at all. Once #776 is merged it will be easier to implement new kinds of checkpointing callbacks. For example, a simple callback which does the save at the end of each training epoch.

**Comment by hjalmarlucius** [comment]
@festeh for training a model for using, I agree. But for testing ideas, the objective is often to just run it until convergence - but as runs may take days you need to prune. Example: You test 20 ideas for 10 epochs each, then take the best three and test them another 10 epochs to see which is better.

@kuynzereb Great that there is more flex on callbacks. I think ModelCheckpoints ideally should be combined with a user-defined function (as for `training_step`) - e.g. receive a name/metrics `dict` of files currently saved and return name/metrics `dict`s of which to save and which to delete, using whatever metrics (or none) you want for that purpose.

**Comment by Ir1d** [comment]
I propose to add an option on whether saving the latest model. With this, we can provide the function of checkpointing both best and latest models.

@williamFalcon

**Comment by Borda** [comment]
@kuynzereb callbacks are refactored now, are interested in sending a PR with this kind of checkpoint?

**Comment by mikerossgithub** [comment]
It might be nice to generalize this. Some example use cases: 
1>  Save the last N checkpoints (e.g. for ensembling). 
2> Save checkpoint every M-th epoch for training curve analysis.

**Comment by collinmccarthy** [comment]
Saving both the latest and the best models is exactly what I'm looking for. When I'm running on a SLURM cluster in low priority mode and may frequently get killed, I would like to resume training with the latest model, but know I always have the best model for testing as well. Thank you in advance whenever something like this gets merged!

**Comment by williamFalcon** [comment]
@collinmccarthy on slurm, lightning auto resubmits your job. for this purpose it creates its own checkpoint and restores from that. 

then you in addition can use the save best to track the best one

**Comment by collinmccarthy** [comment]
@williamFalcon Oh awesome, didn't realize that. Thank you!

**Comment by Ir1d** [comment]
when save_best_k is specified, it seems that only the specified ckpt is stored. I think the latest ckpt is not guaranteed to be saved?

**Comment by philip30** [comment]
@williamFalcon It seems that the autoresubmit feature does not work for me. I tried to follow the sbatch script from the documentation. I submit the job using sbatch. How to make this feature work?

**Comment by Borda** [comment]
@philip30 could you open a separate issue for "autoresubmit feature" and describe it more...
